### PR TITLE
Add template interceptor type definition

### PR DESCRIPTION
**Pull request by optical**
Adds definition for the template interceptor.

See https://github.com/cujojs/rest/blob/master/docs/interceptors.md#module-rest/interceptor/template

### CODE REVIEW

**Comment by optical** [comment]
Yeah I wasn't certain about that - I think its more correct though, so will change.

Also, with the tests the current ones all have the config type set to `any` when calling `wrap`. Should these explicitly use their corresponding config type where appropriate?

```
    .wrap<defaultRequest.Config>(defaultRequest)
    .wrap<hateoas.Config>(hateoas)
    .wrap<location.Config>(location)
```

 And so forth, or is there a reason not to do this?

It'd be nice if these types could always be correctly inferred, but I can't seem to get to work.

**Comment by optical** [comment]
Actually it does seem to infer the types properly most of the time.

With the current incorrect definition it'll infer the config type to be `template.Config` and will error out because of the extra property.
`rest.wrap(template, { params: { extra: 'bang!' } });`

However surprisingly this will compile just fine?
`rest.wrap<template.Config>(template, { params: { extra: 'bang!' } });`

So with the tests I could change them to use inference wherever appropriate, and fill out corresponding stub config too.

Does this sound reasonable?

**Comment by Nemo157** [comment]
Weird that it works with the explicit type... Changing to use inference everywhere would be good, they may have improved the inference since I first added the config checking.

**Comment by Nemo157** [comment]
Ah, looks like it broke on TS1.6 for some reason (486828c97de8c080ac2b610ab22414336803ca45 and #5590). If changing back to inferred types works then that would be good. I do remember having some issues with the inference not always working, but can't remember exactly what they were.

File: rest/rest-tests.ts [hunk 1]
```diff
@@ -117,24 +118,25 @@ var promiseOrResponse = interceptor({
...
+    .wrap(errorCode)
+    .wrap(retry)
+    .wrap(template, { template: 'auth={token}', params: { token: 'hunter2' } })
+    .wrap(timeout)
+    .wrap(jsonp)
+    .wrap(xdomain)
+    .wrap(xhr)
+    .wrap(noop)
+    .wrap(fail)
+    .wrap<KnownConfig>(knownConfig, { prop: 'value' })
```
Comment (optical) [inline rest/rest-tests.ts]: Inference doesn't work here

**Comment by Nemo157** [comment]
:+1:

File: rest/rest-tests.ts (hunk 1 above)
Comment (Nemo157) [inline rest/rest-tests.ts]: Ahh, it's failing to derive the correct type for `knownConfig` as it doesn't have an `init` method in the config it passes to `interceptor`. There's probably some way to change the definition of `rest/interceptor` to allow it to correctly derive it.

**Pull request merged**